crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2490,130 +2447,371 @@ impl<'db> Type<'db> {
+                // signature of the dunder method, but will pass in the type of the object as the
+                // "callable type". That ensures that we get errors like "`X` is not callable"
+                // instead of "`<type of illegal '__call__'>` is not callable".
+                let Some((signatures, boundness)) =
+                    self.dunder_signature(db, Some(callable_ty), "__call__")
+                else {
+                    return Signatures::not_callable(callable_ty, self);
                 };
+                let mut signatures = signatures.clone();
+                signatures.set_dunder_call_boundness(boundness);
+                signatures
+            }
 
-                if let Some(function_ty @ Type::FunctionLiteral(function)) =
-                    arguments.first_argument()
-                {
+            // Dynamic types are callable, and the return type is the same dynamic type. Similarly,
+            // `Never` is always callable and returns `Never`.
+            Type::Dynamic(_) | Type::Never => Signatures::single(CallableSignature::dynamic(self)),
+
+            // Note that this correctly returns `None` if none of the union elements are callable.
+            Type::Union(union) => Signatures::from_union(
+                callable_ty,
+                self,
+                union
+                    .elements(db)
+                    .iter()
+                    .map(|element| element.signatures(db, *element)),
+            ),
+
+            Type::Intersection(_) => {
+                Signatures::single(CallableSignature::todo("Type::Intersection.call()"))
+            }
+
+            _ => Signatures::not_callable(callable_ty, self),
+        }
+    }
+
+    /// Calls `self`. Returns a [`CallError`] if `self` is (always or possibly) not callable, or if
+    /// the arguments are not compatible with the formal parameters. You get back a [`Bindings`]
+    /// for both successful and unsuccessful calls. It contains information about which formal
+    /// parameters each argument was matched to, and about any errors matching arguments and
+    /// parameters.
+    fn try_call(
+        self,
+        db: &'db dyn Db,
+        arguments: &CallArguments<'_, 'db>,
+    ) -> Result<Bindings<'db>, CallError<'db>> {
+        let signatures = self.signatures(db, self);
+        let mut bindings = Bindings::bind(db, signatures, arguments).into_result()?;
+        for binding in bindings.iter_mut() {
+            // For certain known callables, we have special case logic to determine the return type
+            // in a way that isn't directly expressible in the type system. Each special case
+            // listed here should have a corresponding clause above in `signatures`.
+            let Some((overload_index, overload)) = binding.matching_overload_mut() else {
+                continue;
+            };
+
+            match self {
```

Comment:
Not sure, but does this break handling of these special known callables when they occur inside a union? Should we be matching on a `callable_type` or `signature_type` from the binding's specific signature, rather than on `self` here?

Reply:
It sure does, great catch!  Fixed and added an mdtest for this